moshi/src/main/java/com/squareup/moshi/ClassFactory.java

```diff
@@ -98,57 +90,91 @@ public T newInstance() throws InvocationTargetException, IllegalAccessException
+          @Override
+          public T newInstance() throws InvocationTargetException, IllegalAccessException {
+            return (T) newInstance.invoke(null, rawType, constructorId);
+          }
+          @Override public String toString() {
+            return rawType.getName();
+          }
+        };
+      } catch (IllegalAccessException e) {
+        throw new AssertionError();
+      } catch (InvocationTargetException e) {
+        throw Util.rethrowCause(e);
+      } catch (NoSuchMethodException ignored) {
+        // Not the expected version of Dalvik/libcore!
+      }
+
     }
 
-    // Try (pre-Gingerbread) Dalvik/libcore's ObjectInputStream mechanism.
-    // public class ObjectInputStream {
-    //   private static native Object newInstance(
-    //     Class<?> instantiationClass, Class<?> constructorClass);
-    // }
-    try {
-      final Method newInstance = ObjectInputStream.class.getDeclaredMethod(
-          "newInstance", Class.class, Class.class);
-      newInstance.setAccessible(true);
-      return new ClassFactory<T>() {
-        @SuppressWarnings("unchecked")
-        @Override public T newInstance() throws InvocationTargetException, IllegalAccessException {
-          return (T) newInstance.invoke(null, rawType, Object.class);
-        }
-        @Override public String toString() {
-          return rawType.getName();
-        }
-      };
-    } catch (Exception ignored) {
+    if (isBelowAndroidApi(9)) {
+      // Try (pre-Gingerbread) Dalvik/libcore's ObjectInputStream mechanism.
+      // public class ObjectInputStream {
+      //   private static native Object newInstance(
+      //     Class<?> instantiationClass, Class<?> constructorClass);
+      // }
+      try {
+        final Method newInstance = ObjectInputStream.class.getDeclaredMethod(
+            "newInstance", Class.class, Class.class);
+        newInstance.setAccessible(true);
+        return new ClassFactory<T>() {
+          @SuppressWarnings("unchecked")
+          @Override
+          public T newInstance() throws InvocationTargetException, IllegalAccessException {
+            return (T) newInstance.invoke(null, rawType, Object.class);
+          }
+          @Override public String toString() {
+            return rawType.getName();
+          }
+        };
+      } catch (Exception ignored) {
+      }
     }
 
     throw new IllegalArgumentException("cannot construct instances of " + rawType.getName());
   }
+
+  private static boolean isAndroid() {
+    if (!androidChecked) {
+      try {
+        Class<?> androidBuildClass = Class.forName("android.os.Build$VERSION");
+        isAndroid = true;
+        Field sdkIntField = androidBuildClass.getDeclaredField("SDK_INT");
+        androidSdkInt = (int) sdkIntField.get(null);
+      } catch (Exception ignored) {
+      } finally {
+        androidChecked = true;
+      }
+    }
+    return isAndroid;
+  }
+
+  private static boolean isBelowAndroidApi(int target) {
```

Comment:
Minor, but I found this method name confusing - what is "is" here?

Reply:
Cleaned up in 415ffb0